pkg/azuredisk/controllerserver.go

```diff
@@ -517,60 +525,170 @@ func (d *Driver) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
+			if _, visited := rgMap[rg]; visited {
+				continue
+			}
+			rgMap[rg] = true
+		}
 	}
 
-	if start != 0 && start >= len(disks) {
+	if start != 0 && start > len(volSet) {
 		return nil, status.Errorf(codes.Aborted, "ListVolumes starting token(%d) on rg(%s) is greater than total number of volumes", start, d.cloud.ResourceGroup)
 	}
+	resourceGroups := make([]string, len(rgMap))
+	i := 0
+	for rg := range rgMap {
+		resourceGroups[i] = rg
+		i++
+	}
+	sort.Strings(resourceGroups)
 
-	maxEntries := len(disks) - start
-	if req.MaxEntries > 0 && int(req.MaxEntries) < maxEntries {
-		maxEntries = int(req.MaxEntries)
+	// loop through each resourceGroup to get disk lists
+	entries := []*csi.ListVolumesResponse_Entry{}
+	numVisited := 0
+	isCompleteRun, startFound := true, false
+	for _, resourceGroup := range resourceGroups {
+		if !isCompleteRun || (maxEntries > 0 && len(entries) >= maxEntries) {
+			isCompleteRun = false
+			break
+		}
+		localStart := start - numVisited
+		if startFound {
+			localStart = 0
+		}
+		listStatus := d.listVolumesByResourceGroup(ctx, resourceGroup, entries, localStart, maxEntries-len(entries), &volSet)
+		numVisited += listStatus.numVisited
+		if listStatus.err != nil {
+			if status.Code(listStatus.err) == codes.FailedPrecondition {
+				continue
+			}
+			return nil, listStatus.err
+		}
+		startFound = true
+		entries = listStatus.entries
+		isCompleteRun = isCompleteRun && listStatus.isCompleteRun
+	}
+	nextTokenString := ""
+	if !isCompleteRun {
+		nextTokenString = strconv.Itoa(start + numVisited)
 	}
-	pageEnd := start + maxEntries
 
-	vmSet := d.cloud.VMSet
+	listVolumesResp := &csi.ListVolumesResponse{
+		Entries:   entries,
+		NextToken: nextTokenString,
+	}
+
+	return listVolumesResp, nil
+}
+
+// listVolumesInNodeResourceGroup is a helper function for ListVolumes used for when there is no available kubeclient
+func (d *Driver) listVolumesInNodeResourceGroup(ctx context.Context, start, maxEntries int) (*csi.ListVolumesResponse, error) {
 	entries := []*csi.ListVolumesResponse_Entry{}
-	for i := start; i < pageEnd; i++ {
-		d := disks[i]
-		nodeList := []string{}
+	listStatus := d.listVolumesByResourceGroup(ctx, d.cloud.ResourceGroup, entries, start, maxEntries, nil)
+	if listStatus.err != nil {
+		return nil, listStatus.err
+	}
+
+	nextTokenString := ""
+	if !listStatus.isCompleteRun {
+		nextTokenString = strconv.Itoa(listStatus.numVisited)
+	}
+
+	listVolumesResp := &csi.ListVolumesResponse{
+		Entries:   listStatus.entries,
+		NextToken: nextTokenString,
+	}
+
+	return listVolumesResp, nil
+}
+
+// listVolumesByResourceGroup is a helper function that updates the ListVolumeResponse_Entry slice and returns number of total visited volumes, number of volumes that needs to be visited and an error if found
+func (d *Driver) listVolumesByResourceGroup(ctx context.Context, resourceGroup string, entries []*csi.ListVolumesResponse_Entry, start, maxEntries int, volSet *map[string]bool) listVolumeStatus {
```

Comment:
add comments for func parameters, especially `volSet`, and why use `*map[string]bool` (pointer), map in golang is already pass by reference

Reply:
I see, `nil` means it won't filter. In this case, if `volSet` is not nil and is empty, then it should return directly, not necessary to run `DisksClient.ListByResourceGroup`

Reply:
With the current design, `DisksClient.ListByResourceGroup` would need to be called nonetheless to update numVisited because start token refers to an index in a list of all azure disks, not azure csi volumes. But if volSet is not nil and is empty, we should certainly skip the unnecessary loop and return `listVolumeStatus` with updated `numVisited` value.

Reply:
If the start token refers to an index in a list of azure csi disk volumes, we would always have to loop from the start of the disk list returned by `DisksClient.ListByResourceGroup` in `Driver.listVolumesByResourceGroup` in order to find where to actually start adding volume entries. And I am afraid that it would make this process too expensive.

Imagine if there were 1000 azure csi disk volumes and our start token is 900. We have one resource group `rg-1`, and `DisksClient.ListByResourceGroup(ctx, "rg-1")` returns 3000 disks, among which there are 1000 csi volumes. We would then need to iterate from the very start of the list and at worst check 2900 entries to determine where our 900th csi volume is.

On the other hand, if the start token refers to an index in a list of all azure disks, we can start iterating from where the start token points to. Of course, this means that we won't be able to return an error for start token that is too large until we loop through all disks from all resource groups, but the frequency of this happening compared to the abovementioned situation seems to be much lower.